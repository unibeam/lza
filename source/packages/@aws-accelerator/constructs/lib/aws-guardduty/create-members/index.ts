/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { chunkArray, setRetryStrategy } from '@aws-accelerator/utils/lib/common-functions';
import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import {
  GuardDutyClient,
  AccountDetail,
  UpdateOrganizationConfigurationCommand,
  CreateMembersCommand,
  ListMembersCommand,
  DisassociateMembersCommand,
  DeleteMembersCommand,
  ListDetectorsCommand,
  OrganizationFeatureConfiguration,
  BadRequestException,
} from '@aws-sdk/client-guardduty';
import { OrganizationsClient, ListAccountsCommand, ListAccountsCommandOutput } from '@aws-sdk/client-organizations';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';

/**
 * enable-guardduty - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
      StatusCode: number | undefined;
    }
  | undefined
> {
  const region = event.ResourceProperties['region'];
  const partition = event.ResourceProperties['partition'];
  const guardDutyMemberAccountIds: string[] = event.ResourceProperties['guardDutyMemberAccountIds'];
  const enableS3Protection: boolean = event.ResourceProperties['enableS3Protection'] === 'true';
  const enableEksProtection: boolean = event.ResourceProperties['enableEksProtection'] === 'true';
  const autoEnableOrgMembersFlag: boolean = event.ResourceProperties['autoEnableOrgMembers'] === 'true';

  const solutionId = process.env['SOLUTION_ID'];
  const chunkSize = process.env['CHUNK_SIZE'] ? parseInt(process.env['CHUNK_SIZE']) : 50;

  let organizationsClient: OrganizationsClient;
  if (partition === 'aws-us-gov') {
    organizationsClient = new OrganizationsClient({
      region: 'us-gov-west-1',
      customUserAgent: solutionId,
      retryStrategy: setRetryStrategy(),
    });
  } else if (partition === 'aws-cn') {
    organizationsClient = new OrganizationsClient({
      region: 'cn-northwest-1',
      customUserAgent: solutionId,
      retryStrategy: setRetryStrategy(),
    });
  } else {
    organizationsClient = new OrganizationsClient({
      region: 'us-east-1',
      customUserAgent: solutionId,
      retryStrategy: setRetryStrategy(),
    });
  }

  const guardDutyClient = new GuardDutyClient({
    region: region,
    customUserAgent: solutionId,
    retryStrategy: setRetryStrategy(),
  });

  const detectorId = await getDetectorId(guardDutyClient);

  let nextToken: string | undefined = undefined;

  let existingMemberAccountIds: string[] = [];

  let autoEnableOrgMembers: 'ALL' | 'NEW' | 'NONE' = 'ALL';
  if (!autoEnableOrgMembersFlag) {
    autoEnableOrgMembers = 'NONE';
  }

  console.log(`EnableS3Protection: ${enableS3Protection}`);
  console.log(`EnableEksProtection: ${enableEksProtection}`);
  console.log(`autoEnableOrgMembers: ${autoEnableOrgMembers}`);

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const features = getOrganizationFeaturesEnabled(enableS3Protection, enableEksProtection);

      console.log('starting - UpdateOrganizationConfiguration');
      try {
        autoEnableOrgMembersFlag
          ? await updateOrganizationConfiguration(guardDutyClient, detectorId!, autoEnableOrgMembers, features)
          : await updateOrganizationConfiguration(guardDutyClient, detectorId!, autoEnableOrgMembers);

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } catch (e: any) {
        return { Status: 'Failure', StatusCode: e.statusCode };
      }

      console.log('starting - CreateMembersCommand');
      const allAccounts: AccountDetail[] = [];

      do {
        const page = await throttlingBackOff(() =>
          organizationsClient.send(new ListAccountsCommand({ NextToken: nextToken })),
        );

        allAccounts.push(...getMembersToCreate(page, guardDutyMemberAccountIds));
        nextToken = page.NextToken;
      } while (nextToken);

      const chunkedAccountsForCreate = chunkArray(allAccounts, chunkSize);

      for (const accounts of chunkedAccountsForCreate) {
        console.log(`Initiating createMembers request for ${accounts.length} accounts`);
        await throttlingBackOff(() =>
          guardDutyClient.send(new CreateMembersCommand({ DetectorId: detectorId!, AccountDetails: accounts })),
        );
      }

      // Cleanup members removed from deploymentTarget
      if (guardDutyMemberAccountIds.length > 0) {
        console.log('Inititating cleanup of members removed from deploymentTargets');
        existingMemberAccountIds = await getExistingMembers(guardDutyClient, detectorId!);

        const memberAccountIdsToDelete: string[] = [];
        for (const accountId of existingMemberAccountIds) {
          if (!guardDutyMemberAccountIds.includes(accountId)) {
            memberAccountIdsToDelete.push(accountId);
          }
        }
        if (memberAccountIdsToDelete.length > 0) {
          const chunkedAccountsForDelete = chunkArray(memberAccountIdsToDelete, chunkSize);
          await disassociateAndDeleteMembers(guardDutyClient, detectorId!, chunkedAccountsForDelete);
        }
      }

      console.log('Returning Success');
      return { Status: 'Success', StatusCode: 200 };

    case 'Delete':
      const disabledFeatures = getOrganizationFeaturesEnabled(false, false);
      try {
        await updateOrganizationConfiguration(guardDutyClient, detectorId!, 'NONE', disabledFeatures);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } catch (e: any) {
        return { Status: 'Failure', StatusCode: e.statusCode };
      }

      existingMemberAccountIds = await getExistingMembers(guardDutyClient, detectorId!);

      if (existingMemberAccountIds.length > 0) {
        const chunkedAccountsForDelete = chunkArray(existingMemberAccountIds, chunkSize);
        await disassociateAndDeleteMembers(guardDutyClient, detectorId!, chunkedAccountsForDelete);
      }

      return { Status: 'Success', StatusCode: 200 };
  }
}

/**
 * Get accounts details for the GuardDuty members to create
 * @param page
 * @param guardDutyMemberAccountIds
 * @returns
 */
function getMembersToCreate(page: ListAccountsCommandOutput, guardDutyMemberAccountIds: string[]): AccountDetail[] {
  const allAccounts: AccountDetail[] = [];
  for (const account of page.Accounts ?? []) {
    if (guardDutyMemberAccountIds.length > 0) {
      if (guardDutyMemberAccountIds.includes(account.Id!)) {
        allAccounts.push({ AccountId: account.Id!, Email: account.Email! });
      }
    } else {
      allAccounts.push({ AccountId: account.Id!, Email: account.Email! });
    }
  }
  return allAccounts;
}

/**
 * Function to get existing guardduty members
 * @param guardDutyClient
 * @param detectorId
 * @returns string[]
 */
async function getExistingMembers(guardDutyClient: GuardDutyClient, detectorId: string): Promise<string[]> {
  const existingMemberAccountIds: string[] = [];
  let nextToken: string | undefined = undefined;
  do {
    const page = await throttlingBackOff(() =>
      guardDutyClient.send(new ListMembersCommand({ DetectorId: detectorId, NextToken: nextToken })),
    );
    for (const member of page.Members ?? []) {
      existingMemberAccountIds.push(member.AccountId!);
    }
    nextToken = page.NextToken;
  } while (nextToken);

  return existingMemberAccountIds;
}

/**
 * Function to disassociate and delete guardduty members
 * @param guardDutyClient
 * @param chunkedAccountsForDelete
 */
async function disassociateAndDeleteMembers(
  guardDutyClient: GuardDutyClient,
  detectorId: string,
  chunkedAccountsForDelete: string[][],
) {
  for (const existingMemberAccountIdBatch of chunkedAccountsForDelete) {
    console.log(`Initiating disassociateMembers request for ${existingMemberAccountIdBatch.length} accounts`);
    await throttlingBackOff(() =>
      guardDutyClient.send(
        new DisassociateMembersCommand({ AccountIds: existingMemberAccountIdBatch, DetectorId: detectorId }),
      ),
    );

    console.log(`Initiating deleteMembers request for ${existingMemberAccountIdBatch.length} accounts`);
    await throttlingBackOff(() =>
      guardDutyClient.send(
        new DeleteMembersCommand({ AccountIds: existingMemberAccountIdBatch, DetectorId: detectorId }),
      ),
    );
  }
}

function convertBooleanToGuardDutyFormat(flag: boolean) {
  if (flag) {
    return 'NEW';
  } else {
    return 'NONE';
  }
}

function getOrganizationFeaturesEnabled(s3DataEvents: boolean, eksAuditLogs: boolean) {
  const featureList: OrganizationFeatureConfiguration[] = [];

  featureList.push({
    AutoEnable: convertBooleanToGuardDutyFormat(s3DataEvents),
    Name: 'S3_DATA_EVENTS',
  });

  featureList.push({
    AutoEnable: convertBooleanToGuardDutyFormat(eksAuditLogs),
    Name: 'EKS_AUDIT_LOGS',
  });
  return featureList;
}

async function getDetectorId(guardDutyClient: GuardDutyClient): Promise<string | undefined> {
  const response = await throttlingBackOff(() => guardDutyClient.send(new ListDetectorsCommand({})));
  console.log(response);
  return response.DetectorIds!.length === 1 ? response.DetectorIds![0] : undefined;
}

async function updateOrganizationConfiguration(
  guardDutyClient: GuardDutyClient,
  detectorId: string,
  autoEnableOrgMembers: 'ALL' | 'NEW' | 'NONE',
  featureList?: OrganizationFeatureConfiguration[],
) {
  try {
    await guardDutyClient.send(
      new UpdateOrganizationConfigurationCommand({
        AutoEnableOrganizationMembers: autoEnableOrgMembers,
        DetectorId: detectorId!,
        Features: featureList,
      }),
    );
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (e: any) {
    if (e instanceof BadRequestException && featureList) {
      console.log('Retrying with only S3 protection');
      const featureListS3Only = featureList.filter(feat => feat.Name === 'S3_DATA_EVENTS');
      await guardDutyClient.send(
        new UpdateOrganizationConfigurationCommand({
          AutoEnableOrganizationMembers: autoEnableOrgMembers,
          DetectorId: detectorId!,
          Features: featureListS3Only,
        }),
      );
    } else {
      console.log(`Error: ${JSON.stringify(e)}`);
      throw new Error('Failed to update GuardDuty Organization Configuration, check logs for details');
    }
  }
}
