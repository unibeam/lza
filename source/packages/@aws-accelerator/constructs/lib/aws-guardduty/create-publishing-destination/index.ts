/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { setRetryStrategy } from '@aws-accelerator/utils/lib/common-functions';
import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import {
  BadRequestException,
  CreateDetectorCommand,
  CreatePublishingDestinationCommand,
  DeletePublishingDestinationCommand,
  DescribePublishingDestinationCommand,
  GuardDutyClient,
  ListDetectorsCommand,
  ListPublishingDestinationsCommand,
  ListPublishingDestinationsCommandOutput,
  UpdatePublishingDestinationCommand,
} from '@aws-sdk/client-guardduty';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';

interface PublishingDestinationOptions {
  /**
   * Publishing destination ARN
   */
  readonly destinationArn: string;
  /**
   * The export destination type
   */
  readonly exportDestinationType: string;
  /**
   * KMS key ARN
   */
  readonly kmsKeyArn: string;
  /**
   * Override existing configuration
   */
  readonly overrideExisting: boolean;
}

/**
 * enable-guardduty - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
      StatusCode: number | undefined;
    }
  | undefined
> {
  const options = setOptions(event.ResourceProperties);

  const guardDutyClient = new GuardDutyClient({
    retryStrategy: setRetryStrategy(),
    customUserAgent: process.env['SOLUTION_ID'],
  });

  let detectorId = await getDetectorId(guardDutyClient);

  switch (event.RequestType) {
    case 'Create':
      console.log('starting - CreatePublishingDestination');
      if (!detectorId) {
        detectorId = await createDetector(guardDutyClient);
      }

      if (options.overrideExisting) {
        await overrideExistingDestination(guardDutyClient, detectorId, options);
      } else {
        try {
          await throttlingBackOff(() =>
            guardDutyClient.send(
              new CreatePublishingDestinationCommand({
                DetectorId: detectorId,
                DestinationType: options.exportDestinationType,
                DestinationProperties: { DestinationArn: options.destinationArn, KmsKeyArn: options.kmsKeyArn },
              }),
            ),
          );
        } catch (err) {
          if (
            err instanceof BadRequestException &&
            err.message.startsWith('The request failed because a publishingDestination already exists')
          ) {
            console.log('Publishing destination already exists.');
            return { Status: 'Success', StatusCode: 200 };
          }
          throw err;
        }
      }
      return { Status: 'Success', StatusCode: 200 };

    case 'Update':
      console.log('starting - UpdatePublishingDestination');

      // if override is enabled, then follow that workflow
      // Note: override will check destinationArn of s3 and kmsKeyArn. If they are the same no change will be done.
      if (options.overrideExisting) {
        await overrideExistingDestination(guardDutyClient, detectorId!, options);
      } else {
        const updateDestinationId = await getDestinationId(guardDutyClient, detectorId!);

        if (updateDestinationId) {
          await throttlingBackOff(() =>
            guardDutyClient.send(
              new UpdatePublishingDestinationCommand({
                DetectorId: detectorId!,
                DestinationId: updateDestinationId,
                DestinationProperties: { DestinationArn: options.destinationArn, KmsKeyArn: options.kmsKeyArn },
              }),
            ),
          );
        }
      }
      return { Status: 'Success', StatusCode: 200 };

    case 'Delete':
      const deleteDestinationId = await getDestinationId(guardDutyClient, detectorId!);

      if (deleteDestinationId) {
        await throttlingBackOff(() =>
          guardDutyClient.send(
            new DeletePublishingDestinationCommand({
              DetectorId: detectorId!,
              DestinationId: deleteDestinationId,
            }),
          ),
        );
      }

      return { Status: 'Success', StatusCode: 200 };
  }
}

/**
 * Set custom resource options for GuardDuty publishing destination
 * @param resourceProperties { [key: string]: any }
 * @param serviceToken string
 * @returns PublishingDestinationOptions
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function setOptions(resourceProperties: { [key: string]: any }): PublishingDestinationOptions {
  return {
    destinationArn: resourceProperties['destinationArn'],
    exportDestinationType: resourceProperties['exportDestinationType'],
    kmsKeyArn: resourceProperties['kmsKeyArn'],
    overrideExisting: resourceProperties['exportDestinationOverride'] === 'true',
  };
}

/**
 * Returns the detector ID, if one exists
 * @param guardDutyClient GuardDutyClient
 * @returns Promise<string | undefined>
 */
async function getDetectorId(guardDutyClient: GuardDutyClient): Promise<string | undefined> {
  const response = await throttlingBackOff(() => guardDutyClient.send(new ListDetectorsCommand({})));
  console.log(response);
  return response.DetectorIds!.length === 1 ? response.DetectorIds![0] : undefined;
}

/**
 * Create a GuardDuty detector
 * @param guardDutyClient GuardDutyClient
 * @returns Promise<string>
 */
async function createDetector(guardDutyClient: GuardDutyClient): Promise<string> {
  const response = await throttlingBackOff(() => guardDutyClient.send(new CreateDetectorCommand({ Enable: true })));

  if (!response.DetectorId) {
    throw new Error(`Unable to create detector`);
  }
  return response.DetectorId;
}

/**
 * List GuardDuty publishing destinations
 * @param guardDutyClient GuardDutyClient
 * @param detectorId string
 * @returns Promise<ListPublishingDestinationsCommandOutput>
 */
async function getPublishingDestinations(
  guardDutyClient: GuardDutyClient,
  detectorId: string,
): Promise<ListPublishingDestinationsCommandOutput> {
  return await throttlingBackOff(() =>
    guardDutyClient.send(new ListPublishingDestinationsCommand({ DetectorId: detectorId })),
  );
}

/**
 * Get the GuardDuty publishing destination ID
 * @param guardDutyClient GuardDutyClient
 * @param detectorId string
 * @returns Promise<string | undefined>
 */
async function getDestinationId(guardDutyClient: GuardDutyClient, detectorId: string): Promise<string | undefined> {
  try {
    const publishingDestinations = await getPublishingDestinations(guardDutyClient, detectorId);

    if (!publishingDestinations.Destinations) {
      throw new Error(`API did not return destinations`);
    }

    return publishingDestinations.Destinations.length === 1
      ? publishingDestinations.Destinations[0].DestinationId
      : undefined;
  } catch (e) {
    throw new Error(`Unable to retrieve destination ID from ListPublishingDestinations command: ${e}`);
  }
}

/**
 * Override existing publishing destination configuration
 * @param guardDutyClient GuardDutyClient
 * @param detectorId string
 * @param options PublishingDestinationOptions
 */
async function overrideExistingDestination(
  guardDutyClient: GuardDutyClient,
  detectorId: string,
  options: PublishingDestinationOptions,
) {
  // Check account for destination
  const publishingDestinations = await getPublishingDestinations(guardDutyClient, detectorId);
  if (publishingDestinations.Destinations && publishingDestinations.Destinations.length > 0) {
    for (const pubDestination of publishingDestinations.Destinations) {
      // Only s3 is currently possible but leaving this logic in place, in case GuardDuty service adds additional destinations
      if (pubDestination.DestinationType == 'S3') {
        // Get current destination id and find the destination arn
        const pubDestinationResponse = await throttlingBackOff(() =>
          guardDutyClient.send(
            new DescribePublishingDestinationCommand({
              DestinationId: pubDestination.DestinationId,
              DetectorId: detectorId,
            }),
          ),
        );

        if (
          pubDestinationResponse.DestinationProperties &&
          pubDestinationResponse.DestinationProperties.DestinationArn === options.destinationArn &&
          pubDestinationResponse.DestinationProperties.KmsKeyArn === options.kmsKeyArn
        ) {
          // KMS and destination are same as input. So no change is needed.
          console.log('No changes are necessary. Destination Arn and KMS key are the same.');
        } else {
          // kms and destination are not the same. So update destination
          await throttlingBackOff(() =>
            guardDutyClient.send(
              new UpdatePublishingDestinationCommand({
                DetectorId: detectorId,
                DestinationId: pubDestination.DestinationId,
                DestinationProperties: { DestinationArn: options.destinationArn, KmsKeyArn: options.kmsKeyArn },
              }),
            ),
          );
        }
      }
    }
  } else {
    // There are no destinations. Create one
    await throttlingBackOff(() =>
      guardDutyClient.send(
        new CreatePublishingDestinationCommand({
          DetectorId: detectorId,
          DestinationType: options.exportDestinationType,
          DestinationProperties: { DestinationArn: options.destinationArn, KmsKeyArn: options.kmsKeyArn },
        }),
      ),
    );
  }
}
