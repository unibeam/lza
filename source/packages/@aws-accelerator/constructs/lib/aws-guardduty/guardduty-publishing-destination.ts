/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Initialized GuardDutyPublishingDestinationProps properties
 */
export interface GuardDutyPublishingDestinationProps {
  /**
   * Export destination type
   */
  readonly exportDestinationType: string;
  /**
   * Export destination type
   */
  readonly exportDestinationOverride: boolean;
  /**
   * Publishing destination arn
   */
  readonly destinationArn: string;
  /**
   * Publishing destination bucket encryption key
   */
  readonly destinationKmsKey: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly logKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

/**
 * Class - GuardDutyPublishingDestination
 */
export class GuardDutyPublishingDestination extends Construct {
  public readonly id: string = '';

  constructor(scope: Construct, id: string, props: GuardDutyPublishingDestinationProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::GuardDutyCreatePublishingDestinationCommand';

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'create-publishing-destination/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Sid: 'GuardDutyCreatePublishingDestinationCommandTaskGuardDutyActions',
          Effect: 'Allow',
          Action: [
            'guardDuty:CreateDetector',
            'guardDuty:CreatePublishingDestination',
            'guardDuty:DeletePublishingDestination',
            'guardDuty:UpdatePublishingDestination',
            'guardDuty:ListDetectors',
            'guardDuty:ListPublishingDestinations',
            'guardduty:DescribePublishingDestination',
            'iam:CreateServiceLinkedRole',
          ],
          Resource: '*',
        },
        {
          Sid: 'GuardDutyCreateBucketPrefix',
          Effect: 'Allow',
          Action: ['s3:ListBucket', 's3:GetObject'],
          Resource: [props.destinationArn, `${props.destinationArn}/*`],
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        exportDestinationType: props.exportDestinationType,
        exportDestinationOverride: props.exportDestinationOverride,
        destinationArn: props.destinationArn,
        kmsKeyArn: props.destinationKmsKey.keyArn,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.logKmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
