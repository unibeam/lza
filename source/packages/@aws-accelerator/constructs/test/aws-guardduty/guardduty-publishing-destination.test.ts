/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { GuardDutyPublishingDestination } from '../../lib/aws-guardduty/guardduty-publishing-destination';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(GuardDutyPublishingDestination): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new GuardDutyPublishingDestination(stack, 'GuardDutyPublishingDestination', {
  destinationArn: `arn:${stack.partition}:s3:::aws-accelerator-guardduty-${stack.account}-${stack.region}`,
  exportDestinationType: 'S3',
  exportDestinationOverride: true,
  destinationKmsKey: new cdk.aws_kms.Key(stack, 'DestinationKey', {}),
  logKmsKey: new cdk.aws_kms.Key(stack, 'LogKey', {}),
  logRetentionInDays: 3653,
});

/**
 * GuardDutyPublishingDestination construct test
 */
describe('GuardDutyPublishingDestination', () => {
  snapShotTest(testNamePrefix, stack);
});
