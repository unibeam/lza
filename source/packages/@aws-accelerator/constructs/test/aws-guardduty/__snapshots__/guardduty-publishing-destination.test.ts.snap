// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuardDutyPublishingDestination Construct(GuardDutyPublishingDestination):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderHandlerB3AE4CE8": {
      "DependsOn": [
        "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderRoleD01DD26B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderRoleD01DD26B",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderLogGroup118A06DB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "LogKeyE83D0C1B",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderHandlerB3AE4CE8",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderRoleD01DD26B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "guardDuty:CreateDetector",
                    "guardDuty:CreatePublishingDestination",
                    "guardDuty:DeletePublishingDestination",
                    "guardDuty:UpdatePublishingDestination",
                    "guardDuty:ListDetectors",
                    "guardDuty:ListPublishingDestinations",
                    "guardduty:DescribePublishingDestination",
                    "iam:CreateServiceLinkedRole",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "GuardDutyCreatePublishingDestinationCommandTaskGuardDutyActions",
                },
                {
                  "Action": [
                    "s3:ListBucket",
                    "s3:GetObject",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":s3:::aws-accelerator-guardduty-",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          "-",
                          {
                            "Ref": "AWS::Region",
                          },
                        ],
                      ],
                    },
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":s3:::aws-accelerator-guardduty-",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          "-",
                          {
                            "Ref": "AWS::Region",
                          },
                          "/*",
                        ],
                      ],
                    },
                  ],
                  "Sid": "GuardDutyCreateBucketPrefix",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DestinationKeyEE1EE37E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "GuardDutyPublishingDestination52AE4412": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderLogGroup118A06DB",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomGuardDutyCreatePublishingDestinationCommandCustomResourceProviderHandlerB3AE4CE8",
            "Arn",
          ],
        },
        "destinationArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":s3:::aws-accelerator-guardduty-",
              {
                "Ref": "AWS::AccountId",
              },
              "-",
              {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "exportDestinationOverride": true,
        "exportDestinationType": "S3",
        "kmsKeyArn": {
          "Fn::GetAtt": [
            "DestinationKeyEE1EE37E",
            "Arn",
          ],
        },
      },
      "Type": "Custom::GuardDutyCreatePublishingDestinationCommand",
      "UpdateReplacePolicy": "Delete",
    },
    "LogKeyE83D0C1B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
